api-consistency.md

```diff
@@ -0,0 +1,169 @@
+API consistency guidelines
+==========================
+
+Preupgrade Assistant provides two separate APIs for writing modules:
+Python API and Bash API.  These languages are very different, so are
+common patterns and methods of solving programming problems.
+
+Yet, we want to keep both APIs retain similar "look & feel" so that
+module developers and maintainers that will need to switch get as little
+surprise as possible.
+
+In order to help developers of both APIs prevent diverging the behaviors
+too much, this document attempts to set interface guidelines that make
+maximum use of both language's common features.
+
+
+Functions
+---------
+
+ *  Python: all functions live under `preupg` module:
+
+        preupg.myfun()
```

Comment:
Sorry for confusion.  Intention of this document is not to suggest rename of any modules but only to address the question of consistency between the two implementations.  I'm using the "preupg" `preupg`/`preupg_` name here mainly for historic reasons.  I did not realize it will be confusing now that the `preup` has been renamed to `preupg`.

However, I believe we both agree that at least one of them (*common.sh*, *script_api.py*) will need to be renamed anyway.  For this reason, let's keep it as t is, I'll change that as soon as the name is settled.

Regarding the prefix, that only applies to Bash API, to make up for the fact that Bash completely lacks any mechanism to control global namespace.  The idea is:

    # foo.py
    BAR = 'hey'
    def baz:
        pass

can be ported to Bash as:

     # foo.sh
     FOO_BAR = 'hey'
     foo_baz() {
         true
     }

so that following calls are equivalent:

     #!/bin/python
     import foo
     print foo.BAR
     foo.baz()

and

     #!/bin/bash
     .  foo.sh
     echo "$FOO_BAR"
     foo_baz

In our case, the "foo" is <strike>"preupg"</strike> "<to_be_decided>".  What's preserved is that:

 *  (global) constants are upper case,
 *  Bash part has virtually no chance of conflict,
 *  the name mapping is as direct as possible,
 *  all is as brief as possible,

Only thing I'm seeing as "con" for the prefixing is that at this moment things like `get_dist_native_list` would get even longer. However, I believe that should (and can, with proper API design) be addressed on the *right* part of the name, ie. by replacing eg. `s/log_error/error/`, `s/get_dist_native_list/ls_native/`...

Also note that in future users (module developers) should not be exposed to (environment) variables as most, if not all should be replaced by functions -- see my proposal #103 for details.